File: packages/sdk-go/ip_filter.go
package pocketping

import (
	"context"
	"fmt"
	"net"
	"net/http"
	"strings"
	"time"
)

// IpFilterMode defines the filtering mode.
type IpFilterMode string

const (
	IpFilterModeAllowlist IpFilterMode = "allowlist"
	IpFilterModeBlocklist IpFilterMode = "blocklist"
	IpFilterModeBoth      IpFilterMode = "both"
)

// IpFilterReason describes why an IP was allowed or blocked.
type IpFilterReason string

const (
	IpFilterReasonAllowlist      IpFilterReason = "allowlist"
	IpFilterReasonBlocklist      IpFilterReason = "blocklist"
	IpFilterReasonCustom         IpFilterReason = "custom"
	IpFilterReasonNotInAllowlist IpFilterReason = "not_in_allowlist"
	IpFilterReasonDefault        IpFilterReason = "default"
)

// IpFilterLogEvent represents a log event for IP filtering actions.
type IpFilterLogEvent struct {
	Type      string         `json:"type"` // "blocked" or "allowed"
	IP        string         `json:"ip"`
	Reason    IpFilterReason `json:"reason"`
	Path      string         `json:"path"`
	Timestamp time.Time      `json:"timestamp"`
	SessionID string         `json:"sessionId,omitempty"`
}

// IpFilterResult represents the result of an IP filter check.
type IpFilterResult struct {
	Allowed bool
	Reason  IpFilterReason
}

// IpFilterCallback is a custom filter function.
// Return true to allow, false to block, nil to defer to list-based filtering.
type IpFilterCallback func(ip string, requestInfo map[string]interface{}) *bool

// IpFilterConfig holds the configuration for IP filtering.
type IpFilterConfig struct {
	// Enabled enables/disables IP filtering (default: false)
	Enabled bool

	// Mode is the filter mode (default: blocklist)
	Mode IpFilterMode

	// Allowlist contains IPs/CIDRs to allow
	Allowlist []string

	// Blocklist contains IPs/CIDRs to block
	Blocklist []string

	// CustomFilter is an optional custom filter callback
	CustomFilter IpFilterCallback

	// LogBlocked logs blocked requests (default: true)
	LogBlocked bool

	// Logger is a custom logger function
	Logger func(event IpFilterLogEvent)

	// BlockedStatusCode is the HTTP status code for blocked requests (default: 403)
	BlockedStatusCode int

	// BlockedMessage is the response message for blocked requests (default: "Forbidden")
	BlockedMessage string

	// TrustProxy trusts proxy headers (default: true)
	TrustProxy bool

	// ProxyHeaders is the list of headers to check for client IP
	ProxyHeaders []string
}

// DefaultIpFilterConfig returns a default IP filter configuration.
func DefaultIpFilterConfig() *IpFilterConfig {
	return &IpFilterConfig{
		Enabled:           false,
		Mode:              IpFilterModeBlocklist,
		Allowlist:         []string{},
		Blocklist:         []string{},
		LogBlocked:        true,
		BlockedStatusCode: http.StatusForbidden,
		BlockedMessage:    "Forbidden",
		TrustProxy:        true,
		ProxyHeaders:      []string{"Cf-Connecting-Ip", "X-Forwarded-For", "X-Real-Ip"},
	}
}

// ipMatchesCidr checks if an IP matches a CIDR range or exact IP.
func ipMatchesCidr(ipStr, cidr string) bool {
	ip := net.ParseIP(ipStr)
	if ip == nil {
		return false
	}

	// If no CIDR suffix, treat as single IP (/32 or /128)
	if !strings.Contains(cidr, "/") {
		cidrIP := net.ParseIP(cidr)
		if cidrIP == nil {
			return false
		}
		return ip.Equal(cidrIP)
	}

	_, network, err := net.ParseCIDR(cidr)
	if err != nil {
		return false
	}

	return network.Contains(ip)
}

// ipMatchesAny checks if an IP matches any entry in the list.
func ipMatchesAny(ip string, entries []string) bool {
	for _, entry := range entries {
		if ipMatchesCidr(ip, entry) {
			return true
		}
	}
	return false
}

// ShouldAllowIP determines if an IP should be allowed based on the filter config.
func ShouldAllowIP(ip string, config *IpFilterConfig) IpFilterResult {
	if config == nil {
		return IpFilterResult{Allowed: true, Reason: IpFilterReasonDefault}
	}

	mode := config.Mode
	if mode == "" {
		mode = IpFilterModeBlocklist
	}

	allowlist := config.Allowlist
	blocklist := config.Blocklist

	switch mode {
	case IpFilterModeAllowlist:
		// Only allow if in allowlist
		if ipMatchesAny(ip, allowlist) {
			return IpFilterResult{Allowed: true, Reason: IpFilterReasonAllowlist}
		}
		return IpFilterResult{Allowed: false, Reason: IpFilterReasonNotInAllowlist}

	case IpFilterModeBlocklist:
		// Block if in blocklist, allow otherwise
		if ipMatchesAny(ip, blocklist) {
			return IpFilterResult{Allowed: false, Reason: IpFilterReasonBlocklist}
		}
		return IpFilterResult{Allowed: true, Reason: IpFilterReasonDefault}

	case IpFilterModeBoth:
		// Allowlist takes precedence, then check blocklist
		if ipMatchesAny(ip, allowlist) {
			return IpFilterResult{Allowed: true, Reason: IpFilterReasonAllowlist}
		}
		if ipMatchesAny(ip, blocklist) {
			return IpFilterResult{Allowed: false, Reason: IpFilterReasonBlocklist}
		}
		return IpFilterResult{Allowed: true, Reason: IpFilterReasonDefault}

	default:
		return IpFilterResult{Allowed: true, Reason: IpFilterReasonDefault}
	}
}

// CheckIPFilter checks the IP filter with support for custom filter callback.
func CheckIPFilter(ctx context.Context, ip string, config *IpFilterConfig, requestInfo map[string]interface{}) IpFilterResult {
	if config == nil || !config.Enabled {
		return IpFilterResult{Allowed: true, Reason: IpFilterReasonDefault}
	}

	// 1. Check custom filter first
	if config.CustomFilter != nil {
		result := config.CustomFilter(ip, requestInfo)
		if result != nil {
			if *result {
				return IpFilterResult{Allowed: true, Reason: IpFilterReasonCustom}
			}
			return IpFilterResult{Allowed: false, Reason: IpFilterReasonCustom}
		}
		// nil = fall through to list-based filtering
	}

	// 2. Apply list-based filtering
	return ShouldAllowIP(ip, config)
}

// LogIPFilterEvent logs an IP filter event.
func (pp *PocketPing) LogIPFilterEvent(event IpFilterLogEvent) {
	if pp.config.IpFilter == nil {
		return
	}

	if pp.config.IpFilter.Logger != nil {
		pp.config.IpFilter.Logger(event)
	} else if pp.config.IpFilter.LogBlocked {
		fmt.Printf("[PocketPing] IP %s: %s - reason: %s (path: %s)\n",
			event.Type, event.IP, event.Reason, event.Path)
	}
}

// CreateIPFilterLogEvent creates an IP filter log event.
func CreateIPFilterLogEvent(eventType, ip string, reason IpFilterReason, path, sessionID string) IpFilterLogEvent {
	return IpFilterLogEvent{
		Type:      eventType,
		IP:        ip,
		Reason:    reason,
		Path:      path,
		Timestamp: time.Now().UTC(),
		SessionID: sessionID,
	}
}

// CheckIPFilterRequest checks if an HTTP request should be allowed based on IP filter config.
// Returns (allowed, clientIP) - if not allowed, the caller should return a 403 response.
func (pp *PocketPing) CheckIPFilterRequest(r *http.Request) (bool, string) {
	if pp.config.IpFilter == nil || !pp.config.IpFilter.Enabled {
		return true, ""
	}

	clientIP := GetClientIP(r, pp.config.IpFilter)
	result := CheckIPFilter(r.Context(), clientIP, pp.config.IpFilter, map[string]interface{}{
		"path":   r.URL.Path,
		"method": r.Method,
	})

	if !result.Allowed {
		if pp.config.IpFilter.LogBlocked {
			event := CreateIPFilterLogEvent("blocked", clientIP, result.Reason, r.URL.Path, "")
			pp.LogIPFilterEvent(event)
		}
		return false, clientIP
	}

	return true, clientIP
}

// WriteIPFilterBlockedResponse writes a blocked response to the HTTP response writer.
func (pp *PocketPing) WriteIPFilterBlockedResponse(w http.ResponseWriter) {
	statusCode := http.StatusForbidden
	message := "Forbidden"

	if pp.config.IpFilter != nil {
		if pp.config.IpFilter.BlockedStatusCode > 0 {
			statusCode = pp.config.IpFilter.BlockedStatusCode
		}
		if pp.config.IpFilter.BlockedMessage != "" {
			message = pp.config.IpFilter.BlockedMessage
		}
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(statusCode)
	fmt.Fprintf(w, `{"error":"%s"}`, message)
}

// GetClientIP extracts the client IP from an HTTP request.
func GetClientIP(r *http.Request, config *IpFilterConfig) string {
	var headers []string
	if config != nil && len(config.ProxyHeaders) > 0 {
		headers = config.ProxyHeaders
	} else {
		headers = []string{"Cf-Connecting-Ip", "X-Forwarded-For", "X-Real-Ip"}
	}

	trustProxy := config == nil || config.TrustProxy

	if trustProxy {
		for _, header := range headers {
			value := r.Header.Get(header)
			if value != "" {
				// X-Forwarded-For can contain multiple IPs, take the first one
				if strings.Contains(value, ",") {
					value = strings.TrimSpace(strings.Split(value, ",")[0])
				}
				return value
			}
		}
	}

	// Fall back to RemoteAddr
	ip := r.RemoteAddr
	// RemoteAddr might include port, strip it
	if colonIdx := strings.LastIndex(ip, ":"); colonIdx != -1 {
		// Check if it's IPv6 [::1]:port format
		if ip[0] == '[' {
			if bracketIdx := strings.Index(ip, "]"); bracketIdx != -1 {
				ip = ip[1:bracketIdx]
			}
		} else {
			ip = ip[:colonIdx]
		}
	}

	return ip
}

File: packages/sdk-node/src/utils/ip-filter.ts
/**
 * IP Filtering utilities for PocketPing SDK
 * Supports CIDR notation and individual IP addresses
 */

export type IpFilterMode = 'allowlist' | 'blocklist' | 'both';

export interface IpFilterConfig {
  /** Enable/disable IP filtering (default: false) */
  enabled?: boolean;
  /** Filter mode (default: 'blocklist') */
  mode?: IpFilterMode;
  /** IPs/CIDRs to allow (e.g., ['192.168.1.0/24', '10.0.0.1']) */
  allowlist?: string[];
  /** IPs/CIDRs to block (e.g., ['203.0.113.0/24', '198.51.100.50']) */
  blocklist?: string[];
  /** Custom filter callback for advanced logic */
  customFilter?: IpFilterCallback;
  /** Log blocked requests for security auditing (default: true) */
  logBlocked?: boolean;
  /** Custom logger function */
  logger?: (event: IpFilterLogEvent) => void;
  /** HTTP status code for blocked requests (default: 403) */
  blockedStatusCode?: number;
  /** Response message for blocked requests (default: 'Forbidden') */
  blockedMessage?: string;
  /** Trust proxy headers (X-Forwarded-For, etc.) (default: true) */
  trustProxy?: boolean;
  /** Ordered list of headers to check for client IP */
  proxyHeaders?: string[];
}

export interface IpFilterLogEvent {
  type: 'blocked' | 'allowed';
  ip: string;
  reason: 'allowlist' | 'blocklist' | 'custom' | 'not_in_allowlist' | 'default';
  path: string;
  timestamp: Date;
  sessionId?: string;
}

export interface IpFilterResult {
  allowed: boolean;
  reason: 'allowlist' | 'blocklist' | 'custom' | 'not_in_allowlist' | 'default';
}

/**
 * Custom IP filter callback
 * Return true to allow, false to block, undefined to defer to list-based filtering
 */
export type IpFilterCallback = (
  ip: string,
  request: { path: string; sessionId?: string }
) => boolean | undefined | Promise<boolean | undefined>;

/**
 * Parse an IPv4 address to a 32-bit unsigned number
 * Returns null for invalid IPs
 */
export function ipToNumber(ip: string): number | null {
  const parts = ip.split('.');
  if (parts.length !== 4) return null;

  let num = 0;
  for (const part of parts) {
    const n = parseInt(part, 10);
    if (isNaN(n) || n < 0 || n > 255) return null;
    num = (num << 8) | n;
  }
  return num >>> 0; // Convert to unsigned
}

/**
 * Parse CIDR notation to base IP and mask
 * Supports both '192.168.1.0/24' and '192.168.1.1' formats
 */
export function parseCidr(cidr: string): { base: number; mask: number } | null {
  const [ip, bits] = cidr.split('/');
  const base = ipToNumber(ip);
  if (base === null) return null;

  const prefix = bits ? parseInt(bits, 10) : 32;
  if (isNaN(prefix) || prefix < 0 || prefix > 32) return null;

  // Create mask: prefix 1-bits followed by (32-prefix) 0-bits
  const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;

  return { base: (base & mask) >>> 0, mask };
}

/**
 * Check if an IP matches a CIDR range or exact IP
 */
export function ipMatchesCidr(ip: string, cidr: string): boolean {
  const ipNum = ipToNumber(ip);
  if (ipNum === null) return false;

  const parsed = parseCidr(cidr);
  if (!parsed) return false;

  return ((ipNum & parsed.mask) >>> 0) === parsed.base;
}

/**
 * Check if IP matches any entry in a list of IPs/CIDRs
 */
export function ipMatchesAny(ip: string, list: string[]): boolean {
  return list.some((entry) => ipMatchesCidr(ip, entry));
}

/**
 * Main IP filter function - determines if an IP should be allowed
 */
export function shouldAllowIp(
  ip: string,
  config: IpFilterConfig
): IpFilterResult {
  const { mode = 'blocklist', allowlist = [], blocklist = [] } = config;

  switch (mode) {
    case 'allowlist':
      // Only allow if in allowlist
      if (ipMatchesAny(ip, allowlist)) {
        return { allowed: true, reason: 'allowlist' };
      }
      return { allowed: false, reason: 'not_in_allowlist' };

    case 'blocklist':
      // Block if in blocklist, allow otherwise
      if (ipMatchesAny(ip, blocklist)) {
        return { allowed: false, reason: 'blocklist' };
      }
      return { allowed: true, reason: 'default' };

    case 'both':
      // Allowlist takes precedence, then check blocklist
      if (ipMatchesAny(ip, allowlist)) {
        return { allowed: true, reason: 'allowlist' };
      }
      if (ipMatchesAny(ip, blocklist)) {
        return { allowed: false, reason: 'blocklist' };
      }
      return { allowed: true, reason: 'default' };

    default:
      return { allowed: true, reason: 'default' };
  }
}

/**
 * Check IP filter with support for custom filter callback
 */
export async function checkIpFilter(
  ip: string,
  config: IpFilterConfig,
  requestInfo: { path: string; sessionId?: string }
): Promise<IpFilterResult> {
  // 1. Check custom filter first
  if (config.customFilter) {
    const customResult = await config.customFilter(ip, requestInfo);
    if (customResult === true) return { allowed: true, reason: 'custom' };
    if (customResult === false) return { allowed: false, reason: 'custom' };
    // undefined = fall through to list-based filtering
  }

  // 2. Apply list-based filtering
  return shouldAllowIp(ip, config);
}

/**
 * Default proxy headers to check for client IP (in order of preference)
 */
export const DEFAULT_PROXY_HEADERS = [
  'cf-connecting-ip', // Cloudflare
  'x-forwarded-for', // Standard proxy header
  'x-real-ip', // Nginx
  'x-client-ip', // Apache
];
